### ISSUE TITLE
rospy Subscriber does not connect to all the publishers of a topic

### ISSUE THREAD

**Issue by v-lopez** [ros/ros_comm#202]
And only receives messages from the publishers it is actually connected to.

For instance, running "rostopic echo /rosout" only prints a few log messages from a few applications. 

This is the output of rostopic info /rosout

```
Type: rosgraph_msgs/Log

Publishers: 
 * /app1 (http://robot:34067/)
 * /app2 (http://robot:53670/)
 * /app3 (http://robot:50407/)
 * /app4/point_head_action (http://robot:45480/)
 * /app5 (http://robot:43579/)
 * /app6 (http://robot:45164/)
 * /app7 (http://robot2:40501/)
 * /app8 (http://robot:49552/)
 * /app9 (http://robot2:55526/)
 * /app10 (http://robot2:36649/)
 * /app11 (http://robot2:40978/)
 * /app12 (http://robot:35125/)
 * /app13 (http://robot:44950/)
 * /app14 (http://robot:59950/)
 * /app15 (http://robot2:43987/)
 * /app16 (http://robot:50592/)
 * /app17 (http://robot:52106/)
 * /app18 (http://robot:52659/)
 * /app19 (http://robot:50433/)
 * /app20 (http://robot2:49992/)
 * /app21 (http://robot:58786/)
 * /rosapi (http://robot:52104/)
 * /rosbridge_server (http://robot:46198/)
 * /app22 (http://robot:47355/)
 * /app23 (http://robot:52411/)
 * /app24 (http://robot:58590/)
 * /app25 (http://robot:50120/)
 * /app26  (http://robot:44512/)
 * /app27  (http://robot:40736/)
 * /app28  (http://robot:33595/)
 * /app29  (http://robot:41752/)
 * /app30 (http://robot:49830/)
 * /move_base (http://robot:55811/)
 * /environment_server (http://robot:33761/)
 * /app31 (http://robot:50697/)
 * /app32 (http://robot:46382/)
 * /app33 (http://robot:44993/)
 * /app34 (http://robot:53737/)
 * /app35 (http://robot:37633/)
 * /app36 (http://robot:53405/)
 * /app37 (http://robot:33728/)
 * /app38 (http://robot:34157/)
 * /app39 (http://robot:51042/)
 * /app40 (http://robot:34381/)
 * /app41 (http://robot:50393/)
 * /app42 (http://robot:38230/)
 * /app43 (http://robot2:41628/)
 * /app44 (http://robot:42712/)
 * /app45 (http://robot:46302/)
 * /app46 (http://robot:54856/)
 * /app47 (http://robot:49878/)
 * /app48 (http://robot:44169/)
 * /app49 (http://robot:35799/)
 * /app50 (http://robot:60482/)
 * /app51 (http://robot:36333/)
 * /app52 (http://robot:60218/)
 * /app53 (http://robot:52409/)
 * /app55 * /personServer (http://robot:47482/)
 * /app56 (http://robot2:39157/)
 * /app57 (http://robot2:44847/)
 * /rostopic_10309_1364384297563 (http://robot:58735/)

Subscribers: 
 * /rosout (http://robot:50616/)
 * /rostopic_10309_1364384297563 (http://robot:58735/)

```

And here's the output of "rosnode info /rostopic_10309_1364384297563"

```
Node [/rostopic_10309_1364384297563]
Publications: 
 * /rosout [rosgraph_msgs/Log]

Subscriptions: 
 * /rosout [rosgraph_msgs/Log]

Services: None


contacting node http://robot:58735/ ...
Pid: 10309
Connections:
 * topic: /rosout
    * to: /rosout
    * direction: outbound
    * transport: TCPROS
 * topic: /rosout
    * to: /rostopic_10309_1364384297563
    * direction: outbound
    * transport: TCPROS
 * topic: /rosout
    * to: /app49 (http://robot:35799/)
    * direction: inbound
    * transport: TCPROS
 * topic: /rosout
    * to: /rostopic_10309_1364384297563 (http://robot:58735/)
    * direction: inbound
    * transport: TCPROS
```

As can be seen, it only connects to a few.

If I run rostopic echo /rosout again and print "rosnode info", a different set of applications are connected to the node. 

```
--------------------------------------------------------------------------------
Node [/rostopic_18545_1364385713341]
Publications: 
 * /rosout [rosgraph_msgs/Log]

Subscriptions: 
 * /rosout [rosgraph_msgs/Log]

Services: None


contacting node http://robot:55269/ ...
Pid: 18545
Connections:
 * topic: /rosout
    * to: /rosout
    * direction: outbound
    * transport: TCPROS
 * topic: /rosout
    * to: /rostopic_18545_1364385713341
    * direction: outbound
    * transport: TCPROS
 * topic: /rosout
    * to: /app53 (http://robot:52409/)
    * direction: inbound
    * transport: TCPROS
 * topic: /rosout
    * to: /app2 (http://robot:47355/)
    * direction: inbound
    * transport: TCPROS
 * topic: /rosout
    * to: /app13 (http://robot:44950/)
    * direction: inbound
    * transport: TCPROS
 * topic: /rosout
    * to: /app20 (robot2:49992/)
    * direction: inbound
    * transport: TCPROS
 * topic: /rosout
    * to: /rostopic_18545_1364385713341 (http://robot:55269/)
    * direction: inbound
    * transport: TCPROS
```

We have a mix of Python and C++ publishers connected to this topic, and both of them are sometimes connected to the rostopic echo.

Furthermore, this also happens with other topics that have lots of publishers. Such as /diagnostics

**Comment by v-lopez** [comment]
Roscomm version is 

```
ros-fuerte-ros-comm                              1.8.15-0precise-20120908-1755-+0000                 ROS communications-related packages, including core client libraries (roscpp, rospy, roslisp) and graph introspection tools (rostopic, rosnode, rosservice, rosparam).

```

**Comment by dirk-thomas** [comment]
Please provide a small self contained example to reproduce the issue.

**Comment by jbrindza** [comment]
See #208 for how to reproduce the issue.

**Comment by dirk-thomas** [comment]
Fixed by #208.
